### PR TITLE
mux: using channel map as mux configuration

### PR DESCRIPTION
**Pull request by abonislawski**
Previous sof_mux_config will be deleted along with global frame_format and num_channels for mux.
Mux will obtain format data from buffer params.

### CODE REVIEW

File: src/include/ipc/header.h [hunk 1]
```diff
@@ -170,7 +170,7 @@
 #define SOF_IPC_MESSAGE_ID(x)			((x) & 0xffff)
 
 /** Maximum message size for mailbox Tx/Rx */
-#define SOF_IPC_MSG_MAX_SIZE			384
+#define SOF_IPC_MSG_MAX_SIZE			768
```
[the remark this answers is not in the document]
Comment (abonislawski) [reply src/include/ipc/header.h]: Sure, done

**Comment by abonislawski** [comment]
> So in short you just transmit which channels are to be muxed together using the channel map, then the actual config options come from the connected buffers themselves?
> 
> Should be more reliable if I understood this correctly.

Yes but the most important part is to use channel map for configuration. Num channels and frame format is small thing which should be done anyway

File: src/audio/mux/mux.c [hunk 2]
```diff
@@ -28,54 +28,57 @@
...
+			trace_mux_error("mux_set_values() error: configured "
+					"number of  streams (%u) "
+					"exceeds maximum = "
+					META_QUOTE(MUX_MAX_STREAMS),
+					used_streams);
+			return -EINVAL;
+		}
+
+		streams[j].pipeline_id = ch_map->ext_id;
+		streams[j].mask[ch_map->ch_index] = ch_map->ch_mask;
```
[the remark this answers is not in the document]
Comment (abonislawski) [reply src/audio/mux/mux.c]: I think docs description (and examples) will be better than my words :) 

https://github.com/mmaka1/sof-docs/blob/channel-mapping/api/uapi/channel-mapping.rst

[the remark this answers is not in the document]
Comment (lgirdwood) [reply src/audio/mux/mux.c]: @abonislawski sorry I mean comment this code inline. i.e. briefly describe what it is doing at each block otherwise it's hard to follow.

[the remark this answers is not in the document]
Comment (lgirdwood) [reply src/audio/mux/mux.c]: @abonislawski any eta for the comments, this can be merged soon. thanks.

[the remark this answers is not in the document]
Comment (abonislawski) [reply src/audio/mux/mux.c]: @lgirdwood sorry, its done now but please note this cannot be merged before #2216

**Comment by lgirdwood** [comment]
@abonislawski can you check ICL CI report.

**Comment by lgirdwood** [comment]
@abonislawski oh and these is some alignment issue picked up by checkpatch, can you fix.

**Comment by abonislawski** [comment]
@lgirdwood alignment fixed but not sure about CI because its only "Build not found." info

**Comment by lgirdwood** [comment]
@zrombel CI has the build not found error, can you please check as it's blocking merge.

**Comment by zrombel** [comment]
Mux unit test fails for all targets.

**Comment by abonislawski** [comment]
Tests updated with channel map generation, please review this part (more changes than expected) @lgirdwood @jajanusz @paulstelian97

File: src/audio/mux/mux.c [hunk 3]
```diff
@@ -28,54 +29,58 @@
...
-	if (cfg->num_streams > MUX_MAX_STREAMS) {
-		trace_mux_error("mux_set_values() error: configured number of "
-				"streams (%u) exceeds maximum = "
-				META_QUOTE(MUX_MAX_STREAMS),
-				cfg->num_streams);
-		return -EINVAL;
+	struct mux_stream_data streams[MUX_MAX_STREAMS];
+	struct sof_ipc_channel_map *ch_map;
+	uint8_t used_streams = 0;
+	uint8_t i, j;
```
Comment (lgirdwood) [inline src/audio/mux/mux.c]: Btw, any reason for using 8bit maths here ? This can actually evaluate to larger/slower code on RISC processor like xtensa. Can you check this disassembly for any difference to 32 bit, as CC may optimise to 32bit anyway.

File: test/cmocka/src/audio/mux/demux_copy.c [hunk 4]
```diff
@@ -88,31 +88,52 @@ static int setup_group(void **state)
 	return 0;
 }
 
+static size_t get_stream_map_size(struct test_data *td)
+{
+	size_t size = 0;
+	int i, j;
+
+	for (i = 0; i < MUX_MAX_STREAMS; ++i) {
```
Comment (lgirdwood) [inline test/cmocka/src/audio/mux/demux_copy.c]: please place a comment explaining what we are calculating.

File: test/cmocka/src/audio/mux/mux_copy.c [hunk 5]
```diff
@@ -100,31 +100,52 @@ static int setup_group(void **state)
 	return 0;
 }
 
+static size_t get_stream_map_size(struct test_data *td)
+{
+	size_t size = 0;
+	int i, j;
+
+	for (i = 0; i < MUX_MAX_STREAMS; ++i) {
```
Comment (lgirdwood) [inline test/cmocka/src/audio/mux/mux_copy.c]: Looks like common code now, care to deduplicate.

File: src/audio/mux/mux.c (hunk 3 above)
Comment (abonislawski) [reply src/audio/mux/mux.c]: Just because values are small, changed to 32

File: test/cmocka/src/audio/mux/demux_copy.c (hunk 4 above)
Comment (abonislawski) [reply test/cmocka/src/audio/mux/demux_copy.c]: Done

File: test/cmocka/src/audio/mux/mux_copy.c (hunk 5 above)
Comment (abonislawski) [reply test/cmocka/src/audio/mux/mux_copy.c]: Moved to util.h

**Comment by lgirdwood** [comment]
@abonislawski can you ping this PR once the dependency is merged. Thanks.

**Comment by lgirdwood** [comment]
@abonislawski any update ?

**Comment by abonislawski** [comment]
@lgirdwood dependency #2216 is ready to merge but not merged yet, kernel PR is ready to merge too

**Comment by zrombel** [comment]
Due to merge of PR#2303 this PR must be rebased or test results for KD tests with audio format 24b/32b ignored.

**Comment by juimonen** [comment]
@abonislawski you have the topology part for the configuration or are you doing with some new binary blo generator? I made https://github.com/thesofproject/sof/pull/2345 to demonstrate the topology for demux, so I can modify it for this and we would have proper topology/configuration support. 

We would really need some documentation for mux/demux usage, I've been at least 2 times searching for it with @singalsuo because of need to use mux/demux. I only had some pre-historic mail from @akloniex that I've lost on the way...

**Comment by lgirdwood** [comment]
@juimonen care to create feature request for the docs and write them, I think you are the new expert now !

**Comment by juimonen** [comment]
@abonislawski @mmaka1 how is the ch_coeffs array in sof_ipc_channel_map used?
so is it sparse, like if I have masks set in bit 3 and 5, do I index it like ch_coeffs[3] and ch_coeffs[5] or they are in ch_coeffs[0] and ch_coeffs[1] ?

**Comment by abonislawski** [comment]
@juimonen ch_coeffs is a flexible array to be as small as possible, so in your example you will have ch_coeffs[0] and ch_coeffs[1]

**Comment by juimonen** [comment]
@abonislawski ok thanks. this will be "pita" to parse from topology in kernel, but I'll see what I can do :)

**Comment by lgirdwood** [comment]
@juimonen why do we need to parse coeffs in the kernel ? We should just need to find them and pass them to FW at most ?

**Comment by juimonen** [comment]
@lgirdwood So I guess these coeffs would be coming from topology? And if the amount of coeffs varies between 0-32, so this kind of variable amount of tokens is just nasty. You can't easily copy it from tuples arrays to ipc structs. You can do it, but requires much more effort than simple fixed amount of tuples, that map almost directly to ipc.

**Comment by juimonen** [comment]
@lgirdwood and the problem with channel map is even worse as you might have many of those in one wodget. All can have different size. You can't even reserve correct amount of memory for ipc struct before parsing the whole area (to find out total variable size of all channel maps)

**Comment by lgirdwood** [comment]
@juimonen can you remind us of the structures we are passing here by pasting in reply.

**Comment by juimonen** [comment]
@lgirdwood  I'm talking about the structs in src/include/ipc/channel_map.h

`struct sof_ipc_channel_map {
        uint32_t ch_index;
        uint32_t ext_id;
        uint32_t ch_mask;
        uint32_t reserved;
        int32_t ch_coeffs[0];
} __packed;`

and

`struct sof_ipc_stream_map {
        struct sof_ipc_cmd_hdr hdr;
        uint32_t num_ch_map;
        uint32_t reserved[3];
        struct sof_ipc_channel_map ch_map[0];
} __packed;`

So we should parse these from topology. stream_map contains variable amount of channel maps. channel_map contains variable amount of coeffs.

**Comment by lgirdwood** [comment]
@juimonen looks like we need a v2 for sof_ipc_stream_map and sof_ipc_channel_map since we have 2 variable length blobs in the array (making it impossible to easily index)

**Comment by jajanusz** [comment]
Depends on #2216

**Comment by lgirdwood** [comment]
Moving to v1.7

**Comment by mwasko** [comment]
The recommendation is to cancel this PR due to fact that there is no functional use-case for this PR and it is more a refactor how the configuration is passed which include ABI upgrade (do not want to do that if not really needed).
